### PR TITLE
feat(test-utils): mock storage api

### PR DESCRIPTION
**Pull request by freben**
some tests have started needing this now

### CODE REVIEW

File: packages/test-utils/package.json [hunk 1]
```diff
@@ -41,7 +41,8 @@
     "react": "^16.12.0",
     "react-dom": "^16.12.0",
     "react-router": "^6.0.0-alpha.5",
-    "react-router-dom": "^6.0.0-alpha.5"
+    "react-router-dom": "^6.0.0-alpha.5",
+    "zen-observable": "^0.8.15"
```
Comment (Nek) [inline packages/test-utils/package.json:45]: Why this specific Observable library?

Comment (Nek) [reply packages/test-utils/package.json:45]: Oh. Nevermind. I can see we already use it in other places.

**Review by Nek — APPROVED** [review]
💎

File: packages/test-utils/package.json (hunk 1 above)
Comment (Rugvip) [reply packages/test-utils/package.json:45]: Yep, it's a very lightweight one that pretty much just implements the core spec. It's also a dependency of some of our other dependencies so it comes "for free" 😁

File: packages/test-utils/src/testUtils/apis/StorageApi/MockStorageApi.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  static factory = {
+    implements: storageApiRef,
+    deps: {},
+    factory: () => new MockStorageApi(''),
+  };
+
+  private readonly namespace: string;
+  private readonly data: Map<string, any>;
+
+  constructor(namespace: string, data?: Map<string, any>) {
```
Comment (Rugvip) [inline packages/test-utils/src/testUtils/apis/StorageApi/MockStorageApi.ts:35]: Wonder if we could switch this around or smth? Doesn't really make sense to give a namespace when you're supplying the mock data. Would also be convenient if you can give an object instead. Maybe make it private and add `MockStorageApi.create`?

**Review by Rugvip — APPROVED** [review]
👍 , but let's change the way it's constructed

File: packages/test-utils/src/testUtils/mockApiRegistry.ts [hunk 3]
```diff
@@ -15,12 +15,13 @@
...
 
 import { ApiTestRegistry } from '@backstage/core-api';
-import { MockErrorApi } from './apis';
+import { MockErrorApi, MockStorageApi } from './apis';
 
 export function createMockApiRegistry(): ApiTestRegistry {
   const registry = new ApiTestRegistry();
 
   registry.register(MockErrorApi.factory);
+  registry.register(MockStorageApi.factory);
```
Comment (Rugvip) [inline packages/test-utils/src/testUtils/mockApiRegistry.ts:24]: Question unrelated to this PR, what do you think about this as a pattern for all APIs?

File: packages/test-utils/src/testUtils/apis/StorageApi/MockStorageApi.ts (hunk 2 above)
Comment (freben) [reply packages/test-utils/src/testUtils/apis/StorageApi/MockStorageApi.ts:35]: Ah but it's not meant for mock data specifically - it's just used for making sub-buckets, you would normally never give this parameter yourself

File: packages/test-utils/src/testUtils/mockApiRegistry.ts (hunk 3 above)
Comment (freben) [reply packages/test-utils/src/testUtils/mockApiRegistry.ts:24]: Sure, at least as a convention, looks good

File: packages/test-utils/src/testUtils/apis/StorageApi/MockStorageApi.ts (hunk 2 above)
Comment (Rugvip) [reply packages/test-utils/src/testUtils/apis/StorageApi/MockStorageApi.ts:35]: yeah that's why I think it's best to not make it part of the API

Comment (freben) [reply packages/test-utils/src/testUtils/apis/StorageApi/MockStorageApi.ts:35]: Hm yes but this is not a public api as such at all. It's just for use by the automatic wrapper basically. And we need to be able to pass an instance of the store when creating a bucket. How do you propose that it be structured? I would have liked to make the constructor private actually

Comment (Rugvip) [reply packages/test-utils/src/testUtils/apis/StorageApi/MockStorageApi.ts:35]: It should be public API. The APIs in the mock registry is for when you don't care about the mock, when you essentially just want an empty implementation. If you want to test your components interaction with the storage API, the idea is that you'd do this:

```tsx
const storageApi = MockStorageApi.create({ 'ui/MyComponent/my-key/hide-banner': true });

const rendered = renderInTestApp(
  <ApiProvider apiRef={storageApiRef} api={storageApi}>
    <MyComponent storageKey="my-key">
  </ApiProvider>
)

expect(rendered.queryByTestId('banner')).not.toBeInTheDocument();
```

Comment (freben) [reply packages/test-utils/src/testUtils/apis/StorageApi/MockStorageApi.ts]: Yarrp, made it a create method and private constructor instead

**Pull request merged**